File: sdk/python/meerkat_mobkit/_transport.py
"""Persistent subprocess transport for MobKit JSON-RPC."""
from __future__ import annotations

import asyncio
import json
import logging
import math
import os
import subprocess
import threading
import time
from typing import Any, Callable
from uuid import uuid4

_log = logging.getLogger("meerkat_mobkit")

# Stock gateways bound their complete shutdown path to 300 seconds: two
# five-second admission drains, a 285-second runtime cleanup window (long
# enough for an admitted 120-second provider callback and the final
# 120-second lease-release callback), and a five-second stdout drain. The
# advertised 310-second horizon adds response-delivery/process-reap margin.
# It is also the safe fallback for handshake-capable custom gateways which do
# not yet advertise an explicit horizon.
_GATEWAY_SHUTDOWN_GRACE_SECONDS = 310.0
_MAX_GATEWAY_SHUTDOWN_HORIZON_MS = 2_147_483_647
_PROCESS_TERMINATE_GRACE_SECONDS = 5.0
_PROCESS_KILL_GRACE_SECONDS = 5.0
_GATEWAY_SHUTDOWN_METHOD = "mobkit/shutdown"


def _sanitize_for_json(obj: Any) -> Any:
    """Recursively sanitize a value so json.dumps won't fail.

    Non-serializable leaves (callables, custom objects) are converted to
    their string representation so the callback response always reaches Rust.
    """
    if obj is None or isinstance(obj, (bool, int, float, str)):
        return obj
    if isinstance(obj, dict):
        return {str(k): _sanitize_for_json(v) for k, v in obj.items()}
    if isinstance(obj, (list, tuple)):
        return [_sanitize_for_json(v) for v in obj]
    # Fall back to string repr for non-serializable objects (e.g. tool callables)
    try:
        json.dumps(obj)
        return obj
    except (TypeError, ValueError):
        return str(obj)


class PersistentTransport:
    """Long-lived mobkit-rpc subprocess communicating over stdin/stdout JSON-RPC.

    Uses a background reader thread to multiplex responses and callbacks.
    Unlike the per-call subprocess transport, this keeps the process alive
    so mob state persists across calls. stderr is sent to devnull to avoid
    backpressure deadlocks.
    """

    def __init__(
        self,
        gateway_bin: str,
        *,
        env: dict[str, str] | None = None,
        timeout: float = 60.0,
    ):
        self.gateway_bin = gateway_bin
        self._env = {**os.environ, **(env or {})}
        self._process: subprocess.Popen[bytes] | None = None
        self._timeout = timeout
        self._write_lock = threading.Lock()      # protects stdin writes
        self._pending_lock = threading.Lock()     # protects _pending and _results
        self._pending: dict[str, threading.Event] = {}
        self._results: dict[str, Any] = {}
        self._reader_thread: threading.Thread | None = None
        self._callback_handler: Callable | None = None
        self._loop: asyncio.AbstractEventLoop | None = None
        self._stderr_file = None
        self._supports_shutdown_handshake = False
        self._shutdown_horizon_seconds = _GATEWAY_SHUTDOWN_GRACE_SECONDS

    def set_callback_handler(self, handler: Callable) -> None:
        self._callback_handler = handler

    @property
    def request_timeout(self) -> float:
        """Default timeout, in seconds, for one outbound RPC request."""
        return self._timeout

    def start(self) -> None:
        if self._process is not None and self._process.poll() is None:
            return
        # Transport capabilities belong to one gateway process. A restarted
        # child must negotiate them again through mobkit/init.
        self._supports_shutdown_handshake = False
        self._shutdown_horizon_seconds = _GATEWAY_SHUTDOWN_GRACE_SECONDS
        # Capture event loop for async callback dispatch
        try:
            self._loop = asyncio.get_running_loop()
        except RuntimeError:
            self._loop = None
        stderr_target: Any = subprocess.DEVNULL
        stderr_path = self._env.get("MOBKIT_GATEWAY_STDERR_FILE", "").strip()
        if stderr_path:
            self._stderr_file = open(stderr_path, "ab", buffering=0)
            stderr_target = self._stderr_file

        self._process = subprocess.Popen(
            [self.gateway_bin, "--persistent"],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=stderr_target,
            env=self._env,
        )
        self._reader_thread = threading.Thread(
            target=self._reader_loop, daemon=True, name="mobkit-reader"
        )
        self._reader_thread.start()

    def _reader_loop(self) -> None:
        assert self._process is not None and self._process.stdout is not None
        while True:
            line = self._process.stdout.readline()
            if not line:
                # Process closed stdout — store error for all pending callers and wake them
                with self._pending_lock:
                    for msg_id in self._pending:
                        if msg_id not in self._results:
                            self._results[msg_id] = {
                                "error": {"code": -32099, "message": "subprocess died"}
                            }
                    for event in self._pending.values():
                        event.set()
                break
            try:
                msg = json.loads(line.decode("utf-8"))
            except json.JSONDecodeError:
                _log.warning("transport: non-JSON line from subprocess: %s", line[:200])
                continue

            if "method" in msg:
                # Callback or notification FROM Rust
                self._handle_callback(msg)
            elif "id" in msg:
                # Response to a pending request
                msg_id = str(msg["id"])
                with self._pending_lock:
                    event = self._pending.get(msg_id)
                    if event is not None:
                        self._results[msg_id] = msg
                if event is not None:
                    event.set()
                else:
                    # The caller may already have timed out and removed its
                    # pending entry. Do not retain an unclaimable late result.
                    _log.debug(
                        "transport: dropping response for non-pending id=%s",
                        msg_id,
                    )
            else:
                _log.warning(
                    "transport: unrecognized message (no id or method): %s",
                    str(msg)[:200],
                )

    def _handle_callback(self, msg: dict) -> None:
        """Dispatch callback in a separate thread so the reader loop is not blocked."""
        if self._callback_handler is None:
            _log.warning(
                "transport: received callback but no handler registered: %s",
                msg.get("method"),
            )
            return
        # Dispatch in a daemon thread to avoid blocking the reader loop
        t = threading.Thread(
            target=self._dispatch_callback, args=(msg,), daemon=True,
            name="mobkit-callback",
        )
        t.start()

    def _dispatch_callback(self, msg: dict) -> None:
        method = msg.get("method", "")
        params = msg.get("params", {})
        callback_id = msg.get("id")  # None for notifications
        try:
            if self._loop is not None and self._loop.is_running():
                future = asyncio.run_coroutine_threadsafe(
                    self._callback_handler(method, params), self._loop
                )
                result = future.result(timeout=self._timeout)
            else:
                raise RuntimeError(
                    "PersistentTransport: no running event loop for callback dispatch"
                )
            # Notifications (no id) are fire-and-forget — no response sent
            if callback_id is None:
                return
            # Ensure result is JSON-serializable before building response.
            # Tools or other callback results may contain non-serializable objects;
            # sanitize them to strings to prevent json.dumps failures in _write_line.
            response = {"jsonrpc": "2.0", "id": callback_id, "result": _sanitize_for_json(result)}
            self._write_line(response)
        except Exception as exc:
            # Notifications: log only, don't try to send error response
            if callback_id is None:
                _log.warning("notification dispatch error (%s): %s", method, exc)
                return
            _log.warning("callback dispatch error: %s", exc)
            error_response = {
                "jsonrpc": "2.0",
                "id": callback_id,
                "error": {"code": -32000, "message": str(exc)},
            }
            try:
                self._write_line(error_response)
            except Exception:
                _log.error("failed to send callback error response for id=%s", callback_id)

    def _write_line(self, obj: dict) -> None:
        with self._write_lock:
            if self._process and self._process.stdin:
                data = json.dumps(obj) + "\n"
                self._process.stdin.write(data.encode("utf-8"))
                self._process.stdin.flush()

    def send_sync(
        self,
        request: dict[str, Any],
        *,
        timeout: float | None = None,
    ) -> Any:
        self._ensure_running()
        response = self._send_sync_running(request, timeout=timeout)
        if request.get("method") == "mobkit/init":
            result = response.get("result") if isinstance(response, dict) else None
            self._supports_shutdown_handshake = bool(
                isinstance(result, dict)
                and result.get("stdio_shutdown_handshake") is True
            )
            self._shutdown_horizon_seconds = _GATEWAY_SHUTDOWN_GRACE_SECONDS
            if self._supports_shutdown_handshake and isinstance(result, dict):
                horizon_ms = result.get("stdio_shutdown_horizon_ms")
                if (
                    isinstance(horizon_ms, int)
                    and not isinstance(horizon_ms, bool)
                    and 0 < horizon_ms <= _MAX_GATEWAY_SHUTDOWN_HORIZON_MS
                ):
                    self._shutdown_horizon_seconds = horizon_ms / 1000.0
        return response

    def _send_sync_running(
        self,
        request: dict[str, Any],
        *,
        timeout: float | None = None,
    ) -> Any:
        """Send on the current child without starting a replacement process."""
        request_timeout = self._timeout if timeout is None else timeout
        if (
            isinstance(request_timeout, bool)
            or not isinstance(request_timeout, (int, float))
            or not math.isfinite(request_timeout)
            or request_timeout <= 0
        ):
            raise ValueError(
                "persistent transport: timeout must be a positive finite number"
            )
        # Pre-fix, requests with no `id` (or two callers using the
        # same id) collided on `self._pending[""]`: the second
        # `_pending[msg_id] = event` clobbered the first caller's
        # Event, blocking it for the full timeout. Reject either
        # condition explicitly so the deadlock surfaces as a clear
        # ValueError at the call site.
        raw_id = request.get("id")
        if raw_id is None or (isinstance(raw_id, str) and not raw_id):
            raise ValueError(
                "persistent transport: request must carry a non-empty `id` "
                "(use uuid4 or similar) — empty/missing ids collide on the "
                "in-flight pending map and deadlock concurrent callers"
            )
        msg_id = str(raw_id)
        event = threading.Event()
        with self._pending_lock:
            if msg_id in self._pending:
                raise ValueError(
                    f"persistent transport: request id {msg_id!r} is already "
                    f"in flight; concurrent callers must use distinct ids"
                )
            self._pending[msg_id] = event
        # Write request (lock only for write, release before wait)
        self._write_line(request)
        # Wait for response — no locks held
        if not event.wait(timeout=request_timeout):
            with self._pending_lock:
                self._pending.pop(msg_id, None)
                self._results.pop(msg_id, None)
            raise RuntimeError(
                f"persistent transport: timeout after {request_timeout}s "
                "waiting for response"
            )
        with self._pending_lock:
            self._pending.pop(msg_id, None)
            result = self._results.pop(msg_id, None)
        if result is None:
            raise RuntimeError("persistent transport: subprocess closed stdout")
        return result

    def _request_gateway_shutdown(
        self,
        process: subprocess.Popen[bytes],
        *,
        timeout: float,
    ) -> None:
        """Wait for runtime cleanup while the current child's stdin stays open."""
        if self._process is not process or process.poll() is not None:
            raise RuntimeError("gateway exited before shutdown handshake")
        response = self._send_sync_running(
            {
                "jsonrpc": "2.0",
                "id": f"mobkit-shutdown-{uuid4()}",
                "method": _GATEWAY_SHUTDOWN_METHOD,
                "params": {},
            },
            timeout=timeout,
        )
        if not isinstance(response, dict):
            raise RuntimeError("gateway shutdown returned a malformed response")
        error = response.get("error")
        if isinstance(error, dict):
            message = error.get("message", "unknown gateway error")
            raise RuntimeError(f"gateway shutdown failed: {message}")
        result = response.get("result")
        if (
            not isinstance(result, dict)
            or result.get("shutdown") is not True
            or result.get("runtime_cleanup_completed") is False
        ):
            raise RuntimeError(
                "gateway shutdown did not complete runtime-owned cleanup"
            )

    async def send_async(
        self,
        request: dict[str, Any],
        *,
        timeout: float | None = None,
    ) -> Any:
        return await asyncio.to_thread(self.send_sync, request, timeout=timeout)

    def stop(self) -> None:
        process = getattr(self, "_process", None)
        if process is None:
            return
        shutdown_horizon = getattr(
            self,
            "_shutdown_horizon_seconds",
            _GATEWAY_SHUTDOWN_GRACE_SECONDS,
        )
        shutdown_error: Exception | None = None
        shutdown_started = time.monotonic()
        try:
            # The gateway may need to round-trip lease/continuity provider
            # callbacks while UnifiedRuntime shuts down. Keep stdin open until
            # a capable gateway acknowledges that cleanup is complete. Older
            # or custom gateways stay on the EOF protocol below.
            if getattr(self, "_supports_shutdown_handshake", False):
                try:
                    self._request_gateway_shutdown(
                        process,
                        timeout=shutdown_horizon,
                    )
                except Exception as exc:
                    _log.debug("gateway shutdown handshake failed: %s", exc)
                    shutdown_error = exc

            elapsed = time.monotonic() - shutdown_started
            remaining_grace = max(0.0, shutdown_horizon - elapsed)
            if process.stdin:
                try:
                    process.stdin.close()
                except OSError:
                    # A gateway that already closed its pipe still needs to be
                    # reaped below.
                    pass
            try:
                process.wait(timeout=remaining_grace)
            except subprocess.TimeoutExpired:
                try:
                    process.terminate()
                except OSError:
                    # The child can exit between the timed wait and signal.
                    pass
                try:
                    process.wait(timeout=_PROCESS_TERMINATE_GRACE_SECONDS)
                except subprocess.TimeoutExpired:
                    try:
                        process.kill()
                    except OSError:
                        pass
                    try:
                        process.wait(timeout=_PROCESS_KILL_GRACE_SECONDS)
                    except (OSError, subprocess.TimeoutExpired):
                        # Signals are best effort; never make SDK teardown
                        # unbounded if the OS cannot reap the child promptly.
                        pass
            except OSError:
                # Already-reaped children need no further cleanup.
                pass
        finally:
            self._process = None
            stderr_file = getattr(self, "_stderr_file", None)
            if stderr_file is not None:
                stderr_file.close()
                self._stderr_file = None
        if shutdown_error is not None:
            raise RuntimeError(
                "persistent transport: gateway shutdown failed after bounded cleanup"
            ) from shutdown_error

    def is_running(self) -> bool:
        return self._process is not None and self._process.poll() is None

    def _ensure_running(self) -> None:
        if not self.is_running():
            self.start()

    def __del__(self) -> None:
        try:
            self.stop()
        except Exception:
            # Explicit shutdown surfaces handshake failures. Destructors run
            # outside a caller-owned error channel and must stay best effort.
            pass


def create_persistent_transport(gateway_bin: str, **kwargs: Any) -> PersistentTransport:
    transport = PersistentTransport(gateway_bin, **kwargs)
    transport.start()
    return transport

File: sdk/typescript/src/transport.ts
/**
 * Persistent subprocess transport for MobKit JSON-RPC.
 *
 * Keeps a long-lived gateway binary alive, communicating over stdin/stdout
 * newline-delimited JSON. Supports bidirectional callbacks from Rust.
 */

import { spawn, spawnSync } from "node:child_process";
import { randomUUID } from "node:crypto";
import { performance } from "node:perf_hooks";
import { createInterface } from "node:readline";
import type { ChildProcess } from "node:child_process";

// -- Types ----------------------------------------------------------------

export interface JsonRpcRequest {
  readonly jsonrpc: "2.0";
  readonly id: string;
  readonly method: string;
  readonly params: Record<string, unknown>;
}

export interface JsonRpcSuccess {
  readonly jsonrpc: "2.0";
  readonly id: string;
  readonly result: unknown;
}

export interface JsonRpcErrorBody {
  readonly code: number;
  readonly message: string;
}

export interface JsonRpcErrorResponse {
  readonly jsonrpc: "2.0";
  readonly id: string;
  readonly error: JsonRpcErrorBody;
}

export type JsonRpcResponse = JsonRpcSuccess | JsonRpcErrorResponse;

export type JsonRpcTransport = (
  request: JsonRpcRequest,
) => Promise<unknown>;

export type JsonRpcSyncTransport = (request: JsonRpcRequest) => unknown;

export type CallbackHandler = (
  method: string,
  params: Record<string, unknown>,
) => Promise<unknown>;

export type FetchLikeResponse = {
  ok: boolean;
  status: number;
  text(): Promise<string>;
};

export type FetchLike = (
  url: string,
  init: {
    method: "POST";
    headers: Record<string, string>;
    body: string;
    /** Optional AbortSignal — surfaced so the http transport can cancel
     * a hung server request after `timeoutMs`. Implementations that
     * don't support it can ignore the field. */
    signal?: AbortSignal;
  },
) => Promise<FetchLikeResponse>;

/**
 * Grace period for the persistent gateway to finish its own shutdown.
 *
 * Provider operations are publicly required to finish within 120 seconds;
 * the stock Rust gateway gives each callback a hard 130-second wire deadline.
 * Its advertised 335-second horizon covers two such callback windows,
 * runtime event/mob drains, bounded RPC/HTTP/stdout phases, and
 * response-delivery/process-reap margin. The same value is the safe fallback
 * for handshake-capable gateways without the newer explicit capability.
 */
export const PERSISTENT_TRANSPORT_SHUTDOWN_GRACE_MS = 335_000;

const PERSISTENT_TRANSPORT_SIGTERM_GRACE_MS = 5_000;
const PERSISTENT_TRANSPORT_SIGKILL_GRACE_MS = 5_000;
const MAX_GATEWAY_SHUTDOWN_HORIZON_MS = 2_147_483_647;
const GATEWAY_SHUTDOWN_METHOD = "mobkit/shutdown";

type ChildExitWaiter = (
  child: ChildProcess,
  timeoutMs: number,
) => Promise<boolean>;

// -- Helpers --------------------------------------------------------------

export function buildJsonRpcRequest(
  id: string,
  method: string,
  params: Record<string, unknown>,
): JsonRpcRequest {
  return { jsonrpc: "2.0", id, method, params };
}

function sanitizeForJson(obj: unknown): unknown {
  if (obj === null || obj === undefined) return obj;
  if (typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") return obj;
  if (Array.isArray(obj)) return obj.map(sanitizeForJson);
  if (typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
      result[k] = sanitizeForJson(v);
    }
    return result;
  }
  return String(obj);
}

function childHasExited(child: ChildProcess): boolean {
  return child.exitCode !== null || child.signalCode !== null;
}

function validateGatewayShutdownResponse(response: unknown): void {
  if (typeof response !== "object" || response === null) {
    throw new Error("gateway shutdown returned a malformed response");
  }
  const envelope = response as Record<string, unknown>;
  if (typeof envelope.error === "object" && envelope.error !== null) {
    const message = String(
      (envelope.error as Record<string, unknown>).message ?? "unknown gateway error",
    );
    throw new Error(`gateway shutdown failed: ${message}`);
  }
  const result = envelope.result;
  if (
    typeof result !== "object" ||
    result === null ||
    (result as Record<string, unknown>).shutdown !== true ||
    (result as Record<string, unknown>).runtime_cleanup_completed !== true
  ) {
    throw new Error("gateway shutdown did not complete runtime-owned cleanup");
  }
}

async function waitForChildExit(
  child: ChildProcess,
  timeoutMs: number,
): Promise<boolean> {
  if (childHasExited(child)) return true;

  return new Promise<boolean>((resolve) => {
    let settled = false;
    let timer: NodeJS.Timeout | null = null;

    const finish = (exited: boolean): void => {
      if (settled) return;
      settled = true;
      child.removeListener("exit", onExit);
      child.removeListener("close", onExit);
      if (timer !== null) clearTimeout(timer);
      resolve(exited);
    };
    const onExit = (): void => finish(true);

    child.once("exit", onExit);
    child.once("close", onExit);

    // Close the check/listener race: the process may have exited between the
    // initial fast path and listener registration.
    if (childHasExited(child)) {
      finish(true);
      return;
    }

    timer = setTimeout(() => finish(childHasExited(child)), timeoutMs);
    timer.unref?.();
    if (settled) clearTimeout(timer);
  });
}

/** @internal Exported for deterministic lifecycle policy tests. */
export async function stopChildProcess(
  child: ChildProcess,
  waitForExit: ChildExitWaiter = waitForChildExit,
  shutdownGraceMs = PERSISTENT_TRANSPORT_SHUTDOWN_GRACE_MS,
): Promise<void> {
  try {
    child.stdin?.end();
  } catch {
    // Continue with signal-based cleanup if stdin is already unavailable.
  }

  if (
    await waitForExit(child, shutdownGraceMs)
  ) {
    return;
  }

  try {
    child.kill("SIGTERM");
  } catch {
    // A concurrent exit is observed by the bounded wait below.
  }

  if (await waitForExit(child, PERSISTENT_TRANSPORT_SIGTERM_GRACE_MS)) {
    return;
  }

  try {
    child.kill("SIGKILL");
  } catch {
    // Best effort: never make SDK shutdown unbounded on a failed kill call.
  }

  if (!(await waitForExit(child, PERSISTENT_TRANSPORT_SIGKILL_GRACE_MS))) {
    throw new Error(
      "persistent transport: gateway process did not terminate after bounded cleanup",
    );
  }
}

// -- PersistentTransport --------------------------------------------------

/**
 * Long-lived gateway subprocess communicating over stdin/stdout JSON-RPC.
 *
 * Uses a readline reader to multiplex responses and callbacks. Unlike
 * per-call subprocess transports, this keeps the process alive so mob
 * state persists across calls.
 */
export class PersistentTransport {
  private _process: ChildProcess | null = null;
  private _stopping: Promise<void> | null = null;
  private readonly _env: Record<string, string>;
  private readonly _timeout: number;
  private _callbackHandler: CallbackHandler | null = null;
  private _supportsShutdownHandshake = false;
  private _shutdownHorizonMs = PERSISTENT_TRANSPORT_SHUTDOWN_GRACE_MS;
  private readonly _pending = new Map<
    string,
    { resolve: (value: unknown) => void; reject: (error: Error) => void }
  >();

  constructor(
    readonly gatewayBin: string,
    options?: { env?: Record<string, string>; timeout?: number },
  ) {
    this._env = { ...process.env, ...(options?.env ?? {}) } as Record<string, string>;
    this._timeout = options?.timeout ?? 60_000;
  }

  setCallbackHandler(handler: CallbackHandler): void {
    this._callbackHandler = handler;
  }

  start(): void {
    if (this._stopping !== null) {
      throw new Error("persistent transport is stopping");
    }
    if (this._process !== null && this._process.exitCode === null) {
      return;
    }

    // Capabilities are process-scoped and must be renegotiated on init after
    // a child restart.
    this._supportsShutdownHandshake = false;
    this._shutdownHorizonMs = PERSISTENT_TRANSPORT_SHUTDOWN_GRACE_MS;
    this._process = spawn(this.gatewayBin, ["--persistent"], {
      env: this._env,
      stdio: ["pipe", "pipe", "ignore"],
    });

    const child = this._process;

    // Background reader on stdout
    if (child.stdout) {
      const rl = createInterface({ input: child.stdout });
      rl.on("line", (line: string) => {
        let msg: Record<string, unknown>;
        try {
          msg = JSON.parse(line) as Record<string, unknown>;
        } catch {
          return;
        }

        if ("method" in msg) {
          this._handleCallback(msg);
        } else if ("id" in msg) {
          const msgId = String(msg.id);
          const pending = this._pending.get(msgId);
          if (pending) {
            this._pending.delete(msgId);
            pending.resolve(msg);
          }
        }
      });

      rl.on("close", () => {
        // Process closed stdout — fail all pending requests
        for (const [id, pending] of this._pending) {
          this._pending.delete(id);
          pending.resolve({
            jsonrpc: "2.0",
            id,
            error: { code: -32099, message: "subprocess died" },
          });
        }
      });
    }

    child.on("error", () => {
      // Process spawn error — fail all pending
      for (const [id, pending] of this._pending) {
        this._pending.delete(id);
        pending.reject(new Error("gateway process failed to start"));
      }
    });
  }

  private _handleCallback(msg: Record<string, unknown>): void {
    if (!this._callbackHandler) return;

    const method = String(msg.method ?? "");
    const params = (
      typeof msg.params === "object" && msg.params !== null
        ? msg.params
        : {}
    ) as Record<string, unknown>;
    const callbackId = msg.id !== undefined ? String(msg.id) : null;

    this._callbackHandler(method, params)
      .then((result) => {
        if (callbackId === null) return; // Notification — no response
        this._writeLine({
          jsonrpc: "2.0",
          id: callbackId,
          result: sanitizeForJson(result),
        });
      })
      .catch((err: unknown) => {
        if (callbackId === null) return;
        this._writeLine({
          jsonrpc: "2.0",
          id: callbackId,
          error: { code: -32000, message: String(err instanceof Error ? err.message : err) },
        });
      });
  }

  private _writeLine(obj: Record<string, unknown>): void {
    if (this._process?.stdin?.writable) {
      this._process.stdin.write(JSON.stringify(obj) + "\n");
    }
  }

  async sendAsync(request: Record<string, unknown>): Promise<unknown> {
    const response = await this._sendAsyncWithTimeout(request, this._timeout);
    if (request.method === "mobkit/init") {
      const result =
        typeof response === "object" && response !== null
          ? (response as Record<string, unknown>).result
          : null;
      this._supportsShutdownHandshake =
        typeof result === "object" &&
        result !== null &&
        (result as Record<string, unknown>).stdio_shutdown_handshake === true;
      this._shutdownHorizonMs = PERSISTENT_TRANSPORT_SHUTDOWN_GRACE_MS;
      if (this._supportsShutdownHandshake) {
        const horizonMs = (result as Record<string, unknown>).stdio_shutdown_horizon_ms;
        if (
          typeof horizonMs === "number" &&
          Number.isSafeInteger(horizonMs) &&
          horizonMs > 0 &&
          horizonMs <= MAX_GATEWAY_SHUTDOWN_HORIZON_MS
        ) {
          this._shutdownHorizonMs = horizonMs;
        }
      }
    }
    return response;
  }

  private async _sendAsyncWithTimeout(
    request: Record<string, unknown>,
    timeoutMs: number,
    expectedChild?: ChildProcess,
  ): Promise<unknown> {
    if (expectedChild === undefined) {
      this._ensureRunning();
    } else if (this._process !== expectedChild || childHasExited(expectedChild)) {
      throw new Error("persistent transport: subprocess is not running");
    }
    const msgId = String(request.id ?? "");

    return new Promise<unknown>((resolve, reject) => {
      const timer = setTimeout(() => {
        this._pending.delete(msgId);
        reject(new Error(`persistent transport: timeout after ${timeoutMs}ms`));
      }, timeoutMs);

      this._pending.set(msgId, {
        resolve: (value) => {
          clearTimeout(timer);
          resolve(value);
        },
        reject: (error) => {
          clearTimeout(timer);
          reject(error);
        },
      });

      try {
        this._writeLine(request as Record<string, unknown>);
      } catch (error) {
        clearTimeout(timer);
        this._pending.delete(msgId);
        reject(error instanceof Error ? error : new Error(String(error)));
      }
    });
  }

  async stop(): Promise<void> {
    if (this._stopping !== null) return this._stopping;

    const child = this._process;
    if (child === null) return;

    const shutdownStarted = performance.now();
    let childTerminated = false;
    let stopping: Promise<void>;
    stopping = (async () => {
      let shutdownError: Error | null = null;
      try {
        // Keep stdin open while the gateway shuts its runtime down: external
        // lease/continuity providers may still need callback round-trips.
        // Capability negotiation keeps older/custom gateways on their EOF
        // protocol instead of assuming method-not-found behavior.
        if (this._supportsShutdownHandshake) {
          const response = await this._sendAsyncWithTimeout(
            {
              jsonrpc: "2.0",
              id: `mobkit-shutdown-${randomUUID()}`,
              method: GATEWAY_SHUTDOWN_METHOD,
              params: {},
            },
            this._shutdownHorizonMs,
            child,
          );
          validateGatewayShutdownResponse(response);
        }
      } catch (error) {
        shutdownError = error instanceof Error ? error : new Error(String(error));
      }
      const elapsedMs = performance.now() - shutdownStarted;
      const remainingGraceMs = Math.max(
        0,
        this._shutdownHorizonMs - elapsedMs,
      );
      await stopChildProcess(child, waitForChildExit, remainingGraceMs);
      childTerminated = true;
      if (shutdownError !== null) {
        throw new Error(
          `persistent transport: gateway shutdown failed after bounded cleanup: ${shutdownError.message}`,
        );
      }
    })().finally(() => {
      if (
        this._process === child &&
        (childTerminated || childHasExited(child))
      ) {
        this._process = null;
      }
      if (this._stopping === stopping) this._stopping = null;
    });
    this._stopping = stopping;
    return stopping;
  }

  isRunning(): boolean {
    return this._process !== null && this._process.exitCode === null;
  }

  private _ensureRunning(): void {
    if (this._stopping !== null) {
      throw new Error("persistent transport is stopping");
    }
    if (!this.isRunning()) {
      this.start();
    }
  }
}

// -- Per-call transport factories -----------------------------------------

/**
 * Create a synchronous transport that spawns the gateway binary per call.
 */
export function createGatewaySyncTransport(
  gatewayBin: string,
): JsonRpcSyncTransport {
  return (request: JsonRpcRequest): unknown => {
    const requestJson = JSON.stringify(request);
    const out = spawnSync(gatewayBin, [], {
      env: { ...process.env, MOBKIT_RPC_REQUEST: requestJson },
      encoding: "utf8",
    });

    if (out.status !== 0) {
      throw new Error(
        `gateway failed (status=${out.status}): ${String(out.stderr ?? "")}`,
      );
    }

    try {
      return JSON.parse(String(out.stdout ?? "")) as unknown;
    } catch {
      throw new Error("gateway returned non-JSON response");
    }
  };
}

/**
 * Create an async transport that spawns the gateway binary per call.
 */
export function createGatewayAsyncTransport(
  gatewayBin: string,
): JsonRpcTransport {
  return async (request: JsonRpcRequest): Promise<unknown> =>
    new Promise<unknown>((resolve, reject) => {
      const requestJson = JSON.stringify(request);
      const child = spawn(gatewayBin, [], {
        env: { ...process.env, MOBKIT_RPC_REQUEST: requestJson },
        stdio: ["ignore", "pipe", "pipe"],
      });

      let stdout = "";
      let stderr = "";

      if (child.stdout) {
        child.stdout.setEncoding("utf8");
        child.stdout.on("data", (chunk: string) => {
          stdout += chunk;
        });
      }
      if (child.stderr) {
        child.stderr.setEncoding("utf8");
        child.stderr.on("data", (chunk: string) => {
          stderr += chunk;
        });
      }

      child.on("error", (error: Error) => reject(error));

      child.on("close", (code: number | null) => {
        if (code !== 0) {
          reject(new Error(`gateway failed (status=${code}): ${stderr}`));
          return;
        }
        try {
          resolve(JSON.parse(stdout) as unknown);
        } catch {
          reject(new Error("gateway returned non-JSON response"));
        }
      });
    });
}

/**
 * Default request timeout for {@link createJsonRpcHttpTransport}. A
 * server that accepts the connection but never replies would otherwise
 * leak the fetch task forever; pre-fix there was no timeout at all.
 */
export const DEFAULT_HTTP_TRANSPORT_TIMEOUT_MS = 60_000;

/**
 * Create an async HTTP POST transport.
 */
export function createJsonRpcHttpTransport(
  endpoint: string,
  options: {
    headers?: Record<string, string>;
    fetchImpl?: FetchLike;
    /** Maximum time (ms) a single request may stay pending before being
     * aborted. Defaults to 60s. Set to 0 to disable (not recommended). */
    timeoutMs?: number;
  } = {},
): JsonRpcTransport {
  const globalFetch = (globalThis as unknown as { fetch?: FetchLike }).fetch;
  const fetchImpl = options.fetchImpl ?? globalFetch;
  if (!fetchImpl) {
    throw new Error("fetch implementation not available");
  }
  const timeoutMs =
    options.timeoutMs === undefined
      ? DEFAULT_HTTP_TRANSPORT_TIMEOUT_MS
      : options.timeoutMs;

  return async (request: JsonRpcRequest): Promise<unknown> => {
    const controller =
      timeoutMs > 0 ? new AbortController() : undefined;
    const timer =
      controller !== undefined && timeoutMs > 0
        ? setTimeout(() => controller.abort(), timeoutMs)
        : undefined;

    try {
      const response = await fetchImpl(endpoint, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          accept: "application/json",
          ...(options.headers ?? {}),
        },
        body: JSON.stringify(request),
        signal: controller?.signal,
      });

      const body = await response.text();
      if (!response.ok) {
        throw new Error(
          `http transport failed (status=${response.status}): ${body}`,
        );
      }

      try {
        return JSON.parse(body) as unknown;
      } catch {
        throw new Error("http transport returned non-JSON response");
      }
    } catch (err) {
      if (
        err !== null &&
        typeof err === "object" &&
        "name" in err &&
        (err as { name: unknown }).name === "AbortError"
      ) {
        throw new Error(
          `http transport timed out after ${timeoutMs}ms; server unresponsive`,
        );
      }
      throw err;
    } finally {
      if (timer !== undefined) {
        clearTimeout(timer);
      }
    }
  };
}
